internal/luks2/metadata.go

```diff
@@ -0,0 +1,874 @@
+)
+
+func (h Hash) GetHash() crypto.Hash {
+	switch h {
+	case HashSHA1:
+		return crypto.SHA1
+	case HashSHA224:
+		return crypto.SHA224
+	case HashSHA256:
+		return crypto.SHA256
+	case HashSHA384:
+		return crypto.SHA384
+	case HashSHA512:
+		return crypto.SHA512
+	default:
+		return 0
+	}
+}
+
+// KeyslotType corresponds to the type of a keyslot.
+type KeyslotType string
+
+const (
+	KeyslotTypeLUKS2 KeyslotType = "luks2"
+)
+
+// AFType corresponds to an anti-forensic splitter algorithm.
+type AFType string
+
+const (
+	AFTypeLUKS1 AFType = "luks1"
+)
+
+// AreaType corresponds to the type of a storage area in the binary
+// keyslots area.
+type AreaType string
+
+const (
+	AreaTypeRaw AreaType = "raw"
+)
+
+type label [48]byte
+
+func (l label) String() string {
+	return strings.TrimRight(string(l[:]), "\x00")
+}
+
+type csumAlg [32]byte
+
+func (a csumAlg) GetHash() crypto.Hash {
+	return Hash(strings.TrimRight(string(a[:]), "\x00")).GetHash()
+}
+
+type binaryHdr struct {
+	Magic       [6]byte
+	Version     uint16
+	HdrSize     uint64
+	SeqId       uint64
+	Label       label
+	CsumAlg     csumAlg
+	Salt        [64]byte
+	Uuid        [40]byte
+	Subsystem   [48]byte
+	HdrOffset   uint64
+	Padding     [184]byte
+	Csum        [64]byte
+	Padding4096 [7 * 512]byte
+}
+
+type jsonNumber string
+
+func (n jsonNumber) int() (int, error) {
+	return strconv.Atoi(string(n))
+}
+
+func (n jsonNumber) uint64() (uint64, error) {
+	return strconv.ParseUint(string(n), 10, 64)
+}
+
+// Config corresponds to a config object in the JSON metadata of a LUKS2 volume.
+type Config struct {
+	JSONSize     uint64   // Size of the JSON area, in bytes
+	KeyslotsSize uint64   // Size of the keyslots area, in bytes
+	Flags        []string // Optional flags
+	Requirements []string // Optional required features
+}
+
+func (c *Config) UnmarshalJSON(data []byte) error {
+	var d struct {
+		JSONSize     jsonNumber `json:"json_size"`
+		KeyslotsSize jsonNumber `json:"keyslots_size"`
+		Flags        []string
+		Requirements []string
+	}
+	if err := json.Unmarshal(data, &d); err != nil {
+		return err
+	}
+
+	*c = Config{
+		Flags:        d.Flags,
+		Requirements: d.Requirements}
+	jsonSize, err := d.JSONSize.uint64()
+	if err != nil {
+		return xerrors.Errorf("invalid json_size value: %w", err)
+	}
+	c.JSONSize = jsonSize
+
+	keyslotsSize, err := d.KeyslotsSize.uint64()
+	if err != nil {
+		return xerrors.Errorf("invalid keyslots_size value: %w", err)
+	}
+	c.KeyslotsSize = keyslotsSize
+
+	return nil
+}
+
+// Token corresponds to a token object in the JSON metadata of a LUKS2 volume. It
+// describes how to retrieve a passphrase or key for a keyslot.
+type Token struct {
+	Type     string                 // Token type ("luks2-" prefixed types are reserved for cryptsetup)
+	Keyslots []int                  // Keyslots assigned to this token
+	Params   map[string]interface{} // Type-specific parameters for this token
+}
+
+func (t Token) MarshalJSON() ([]byte, error) {
```

Comment:
this marshal doesn't seem to be tested

Reply:
There's no explicit test for it in metadata_test.go, but it should get test coverage as part of the ImportToken tests, although I can add some explicit tests too.

Reply:
There's a test for this now.